### ISSUE TITLE
bug(cdk/table): CdkTable does not handle column updates correctly when using recycleRows

### ISSUE THREAD

**Issue by mitchellwills** [angular/components#32253]
### Is this a regression?

- [ ] Yes, this behavior used to work in the previous version

### Description

When using `CdkTable` with the table `recycleRows` directive it doesn't look like the table doesn't re-render cached rows not currently in the DOM. This leads to a situation where if those cached rows are later re-used the cells in the row don't match the columns.

It seems like this may be just a matter of clearing the row cache in `_forceRenderDataRows`. This assumes the row elements can't be re-used at all, but that seems to be the current behavior of `_forceRenderDataRows` to re-render everything. It looks like this could be achieved by calling `detach` on the `_RecycleViewRepeaterStrategy` instance in `_forceRenderDataRows`, but it's unclear the semantics of that method since it's not documented in `_ViewRepeater`. `CdkVirtualForOf` only calls it in `ngOnDestroy` and **`CdkTable` never calls it** (probably a separate bug where these row views get leaked when the table gets destroyed).

### Reproduction

StackBlitz link: https://stackblitz.com/edit/mssn6jgh?file=src%2Fexample%2Fcdk-table-recycle-rows-example.ts
Steps to reproduce:
1. Click `Set Short List` to reduce the number of rows to 15 (caching the 10 removed rows).
2. Then click `Toggle Columns`. This forces a full re-render through `_forceRenderDataRows` (which doesn't populate the cache when removing).

### Expected Behavior

All 15 rows re-render correctly

### Actual Behavior

When the 15 rows are re-rendered 5 have the old list of 3 columns since they were inserted from the cache and the remaining 10 are new rows and render correctly.

### Environment

- Angular: 20.2.4
- CDK/Material: 20.2.4
- Browser(s): 141.0.7390.126
- Operating System (e.g. Windows, macOS, Ubuntu): ChromeOS

**Comment by teja2** [comment]
fix for this issue: please take a look: https://github.com/angular/components/pull/32405

**Comment by ssc-io** [comment]
We just hit this on `@angular/cdk` 22.0.5 (Angular 22.0.7), and I can confirm the diagnosis. Four
additions that may raise the impact of this issue:

**1. It is not limited to opting in via `recycleRows` — it affects every `cdk-table` inside a
`cdk-virtual-scroll-viewport`, and there is no way to opt out.** `ngAfterContentInit` picks the
repeater with an OR:

```ts
this._viewRepeater = this.recycleRows || this._virtualScrollEnabled()
  ? new _RecycleViewRepeaterStrategy()
  : new _DisposeViewRepeaterStrategy();
```

So any virtual-scrolled table whose columns can be shown/hidden or reordered is affected, without
ever touching `recycleRows` (and setting `recycleRows="false"` does not help — it can only turn
recycling *on*). In that setup the cache does not even need a data change to fill up: scrolling
down and back is enough, because rows leaving the rendered range are detached and cached.

**2. The exact mechanism, which confirms the suspected `detach()` fix.** `_forceRenderDataRows()`
resets `_dataDiffer` and clears `_rowOutlet.viewContainer`, but leaves the repeater's `_viewCache`
untouched. In the following `renderRows()` every record looks new (`record.previousIndex == null`),
so `_RecycleViewRepeaterStrategy._insertView()` pops a cached view, inserts it, and returns
`undefined`. Because no view is returned, `applyChanges()` reports the operation as
`_ViewRepeaterOperation.REPLACED` — and the `itemViewChanged` callback in `CdkTable.renderRows()`
stamps cell templates only for `_ViewRepeaterOperation.INSERTED`:

```ts
change => {
  if (change.operation === _ViewRepeaterOperation.INSERTED && change.context) {
    this._renderCellTemplateForItem(change.record.item.rowDef, change.context);
  }
}
```

Hence the recycled rows keep the cells stamped for the previous column list. Header and footer rows
are unaffected because `_forceRenderHeaderRows()` / `_forceRenderFooterRows()` recreate them
directly, which is why header and body end up disagreeing.

**3. Reordering columns is affected as well**, not just showing/hiding: with a filled cache, the
header order updates while the recycled rows keep their cells in the previous order.

**4. Workaround** for anyone blocked by this — purge the cache before the CDK re-renders. A
component effect on the displayed column names runs after the template has updated the row
definition's columns but before the table's `ngAfterContentChecked` acts on the change:

```ts
const repeater = (cdkTable as unknown as Record<string, { detach(): void } | undefined>)['_viewRepeater'];
repeater?.detach();
```

Reaching into a private field is obviously not something we want to keep shipping, hence the
interest in a fix upstream. Both suggested directions look right to me: `this._viewRepeater.detach()`
in `_forceRenderDataRows()` next to the existing `_dataDiffer.diff([])` / `viewContainer.clear()`,
or stamping the cell templates for `REPLACED` views too.
